### ISSUE TITLE
Avoid any commits with files larger than 100KB

### ISSUE THREAD

**Issue by robnagler** [radiasoft/zgoubi#32]
The repo doubled in size to 20MB with 88de5e51fb996e9fd63df0404ecbe1b2fa5e473a and aaa6e072039e6aaa301c1ba7f114c10ee2de5ff2. This was unfortunate, and may have to be undone with another hard reset.

GitHub has been throttling and problematic in general over the last week so the effect is even more pronounced. It took me about 5 minutes to pull a fresh repo yesterday.

We have to be very careful about reasoning about commits with such big data files. The logic for committing warmSnake.map.xz was "replaces warmSnake.map with a compressed version
of same." Unfortunately, this had the opposite effect of what was intended. A git repo is already compressed so adding the xz version changed the repo basically has two versions of warmSnake.map: one compressed by Git and one explicitly compressed. Git doesn't know how to reconcile those two files so the repo now is double in size. 

warmSnake.map does not belong in the repo until we are sure that it makes sense as a unit test support file and can't be trimmed. It's important that we think carefully about commits of any file over 100K. In the mean time, let's add this to the test script so we have an example of the right way to add big files from the old repo:

```sh
if [[ ! -r warmSnake.map ]]; then
    curl -L -S -s https://sourceforge.net/p/zgoubi/code/HEAD/tree/trunk/exemples/usersGuide/FIT-and-REBELOTE/warmSnake.map.gz?format=raw \
        | gunzip -c > warmSnake.map
    [[ $(sum warmSnake.map) == '26951 42103' ]]
fi
```

Once we have a refined set of unit tests, we can start committing those files. For now, they are being used unmodified and can be pulled dynamically.

**Comment by dtabell** [comment]
This is really mysterious to me, because Damian and I discussed this. We compressed the file, so that only warmSnake.map.xz would go into the repo. And in fact that's what I see.

```
09:31:16 dabell@rascal zgoubi-rs> git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
09:31:21 dabell@rascal zgoubi-rs> du -sh tests
5.5M	tests
09:31:37 dabell@rascal zgoubi-rs> ll tests
total 11200
-rw-r--r--  1 dabell  staff   7.7K Jul 18 17:48 PARIZ_cmake_configurable.H
-rw-r--r--  1 dabell  staff   168K Jul 18 17:48 centeredHelix.eps
-rw-r--r--  1 dabell  staff   102K Jul 18 17:48 centeredHelix_FIT_save.res_150226
-rw-r--r--  1 dabell  staff    54K Jul 18 17:48 centeredHelix_FIT_save_nofinal_150226.res
-rw-r--r--  1 dabell  staff   5.1M Jul 18 17:48 warmSnake.map.xz
-rwxr-xr-x  1 dabell  staff   1.1K Jul 18 17:48 zgoubi-integration-test.sh.template*
-rw-r--r--  1 dabell  staff   435B Jul 17 16:55 zgoubi-test.patch
-rw-r--r--  1 dabell  staff    60K Jul 17 16:55 zgoubi.res.expected
```

Just to make sure,

```
09:35:57 dabell@rascal zgoubi-rs> git remote -v
origin	https://github.com/radiasoft/zgoubi (fetch)
origin	https://github.com/radiasoft/zgoubi (push)
```

so this _does_ seem to be correct repository.

**Comment by robnagler** [comment]
The repo is distinct from the copy you check out. Do this:

```sh
$ cd ~/src/radiasoft/zgoubi
$ find .git -size +1M
.git/objects/8a/94c420dea97079adfa489b0d67b52db0a9c490
.git/objects/cc/eac36c34d1131c2ea98dadb247d2c3db2e55bc
.git/objects/pack/pack-b88a5a14f4b034e7f2742e656c0aac3b2008c717.pack
```

Now you can see two blobs with ids `8a94c420dea97079adfa489b0d67b52db0a9c490` and  `cceac36c34d1131c2ea98dadb247d2c3db2e55bc`. You can now:

```sh
$ git cat-file -p 8a94c420dea97079adfa489b0d67b52db0a9c490 | sum
26951 42103
$ git cat-file -p cceac36c34d1131c2ea98dadb247d2c3db2e55bc | xz -d | sum
26951 42103
```

As you can see, these two objects are identical. One is compressed with xz and the other is uncompressed. Their sizes are:

```sh
$ ls -sh -1 .git/objects/8a/94c420dea97079adfa489b0d67b52db0a9c490 .git/objects/cc/eac36c34d1131c2ea98dadb247d2c3db2e55bc
 13M .git/objects/8a/94c420dea97079adfa489b0d67b52db0a9c490
5.1M .git/objects/cc/eac36c34d1131c2ea98dadb247d2c3db2e55bc
```

I was actually wrong in my original statement. The repo was a comfortable 2MB compressed before these two commits. Now it is 10x larger, not 2x larger.

**Comment by rouson** [comment]
I have added this issue to the SOW 1 milestone, which I'm using to track all outstanding issues for SOW 1.  I hope I've caught them all now, but I'll add more if I find that I missed any open issues.

**Comment by zbeekman** [comment]
Unless anyone objects, the download of external data can be accomplished with CMake. This will be portable (whether or not we care about portability) and prevent any dependency on curl, wget or the shell.

@robnagler please let me know if this is an acceptable route for you.

**Comment by robnagler** [comment]
Please proceed with CMake with the following goals:

* The download should only happen if the file doesn't already exist.
* Assert the checksum every build (unlike my proposed snippet which only asserted after download)

If possible, it would be good to build up a structure similar to #29 for storing the copy and running the test. The goals are:

* To have the downloaded data in a separate directory, which gets copied to a run directory when a test is executed.
* When a test fails, the run directory is intact so output can be investigated.
* It should be obvious how to run the test manually so you can debug the failed case.

**Comment by robnagler** [comment]
I assume the procedure is understood now, and we don't need to keep this Issue open.

**Comment by zbeekman** [comment]
@robnagler: Yes we understand, however I have yet to finish implementing the download of the test data via CMake, and was using this issue and your [previous comment](https://github.com/radiasoft/zgoubi/issues/32#issuecomment-407436390) to track that. Can we re-open this issue until I finish?

Also since I have push access, I'd be happy to cleanup the repository using the bfg-repo-cleaner program following the procedure outlined here: https://rtyley.github.io/bfg-repo-cleaner/#usage. If you would like for me to do this just let me know, and also we should confirm that no one else is working from the repository since it would re-write history.

**Comment by robnagler** [comment]
@zbeekman Sorry. I thought that was done in the process of getting CI going.

Happy if you take on the repo cleanup. I won't touch anything. (Technically, I'm on vacation. :)

**Comment by zbeekman** [comment]
@robnagler I was looking through the repo, and there are some source files and result/input files larger than 100kb:

- `centeredHelix_FIT_save.res_150226`: 104K (Although this seems to be safe to delete from the repo AFAICT
- `./zgoubi/svd/DGESVD.f`: 136K
- `./zgoubi/svd/LAPACK.f`: 496K

`DGESVD.f` is currently getting compiled and linked into the `zgoubi` executable, however, `LAPACK.f` is not and appears to be delete-able. 100K is a pretty stringent requirement, which clearly cannot actually be met, at least not until the code is refactored and BLAS/LAPACK can be linked in from implementations provided by package management or as an external third party library.

@robnagler: How should I proceed concerning the repo cleanup?
@dtabell: Do you know if/why we would need `./zgoubi/svd/LAPACK.f` and `centeredHelix_FIT_save.res_150226`? Also, it appears that functionality of `DGESVD.f` can be obtained from LAPACK.

**Comment by robnagler** [comment]
The title is "avoid" so it's "best effort". 

If LAPACK.f is unused, leave it out. Worst case is we have to put it back in.

Leave DGESVD.f in.

**Comment by zbeekman** [comment]
Here is the output/log from running [BFG](https://rtyley.github.io/bfg-repo-cleaner/):

```
$ bfg --strip-blobs-bigger-than 150K zgoubi.git

Using repo : /Users/ibeekman/Sandbox/zgoubi.git

Scanning packfile for large blobs: 16434
Scanning packfile for large blobs completed in 385 ms.
Found 948 blob ids for large blobs - biggest=72135669 smallest=154337
Total size (unpacked)=1921266332
Found 1030 objects to protect
Found 9 commit-pointing refs : HEAD, refs/heads/fix-issue-#28-fedora-27-build, refs/heads/master, ...

Protected commits
-----------------

These are your protected commits, and so their contents will NOT be altered:

 * commit b0eb9994 (protected by 'HEAD')

Cleaning
--------

Found 1068 commits
Cleaning commits:        39% ( 417/1068)
[1]+  Stopped                 bfg --strip-blobs-bigger-than 150K zgoubi.git
Cleaning commits:       100% (1068/1068)
Cleaning commits completed in 7,133 ms.

Updating 8 Refs
---------------

	Ref                                        Before     After
	--------------------------------------------------------------
	refs/heads/fix-issue-#28-fedora-27-build | 270cd30e | 19950fee
	refs/heads/master                        | b0eb9994 | 7520d98b
	refs/pull/22/head                        | 28a06e29 | 2ad94b7c
	refs/pull/23/head                        | 4f0eacd0 | d06621aa
	refs/pull/24/head                        | 0e99731f | 4cf58204
	refs/pull/31/head                        | ddd8cd86 | 05997bbf
	refs/pull/33/head                        | 533b61bd | e35f354a
	refs/pull/33/merge                       | 0e512ddb | a783bfe6

Updating references:    100% (8/8)
...Ref update completed in 191 ms.

Commit Tree-Dirt History
------------------------

	Earliest                                              Latest
	|                                                          |
	DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDD

	D = dirty commits (file tree fixed)
	m = modified commits (commit message or parents changed)
	. = clean commits (no changes to file tree)

	                        Before     After
	-------------------------------------------
	First modified commit | 93663ad6 | 8b724b81
	Last dirty commit     | 2ba7a0a9 | 745f7aad

Deleted files
-------------

	Filename                                                                                                                         Git id
	----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
	08-10-29_Ring5Xtractions-3D.eps                                                                                                | 8e0fbc50 (329.3 KB)
	1-s2.0-S0895611100000562-main.pdf                                                                                              | a095c281 (709.6 KB)
	118-1825_IMG.eps                                                                                                               | 1edccd3e (4.8 MB)
	20GeV_initial_dist                                                                                                             | dcb574c9 (888.9 KB)
	22047216.pdf                                                                                                                   | d88e5849 (167.4 KB)
	23001796.pdf                                                                                                                   | f603e6b4 (161.5 KB)
	2wayFFAG.eps                                                                                                                   | 00e77bcf (2.0 MB)
	5InchCycloHeldByGlenSeaborg.eps                                                                                                | eda59e6f (496.5 KB)
	700MeVFFAG.eps                                                                                                                 | 7221bae6 (964.8 KB)
	85MeVK2.5xi52-pi2.eps                                                                                                          | 932e4bb4 (1.1 MB)
	96hera_future.pdf                                                                                                              | 6b54ef48 (183.8 KB)
	96krisch_part.pdf                                                                                                              | d81818b4 (230.5 KB)
	AD_0360.mpa                                                                                                                    | 723180f7 (629.5 KB)
	AD_1000.mpa                                                                                                                    | dab09487 (629.5 KB)
	AD_2650.mpa                                                                                                                    | f826cfc5 (629.5 KB)
	...


In total, 3451 object ids were changed. Full details are logged here:

	/Users/ibeekman/Sandbox/zgoubi.git.bfg-report/2018-08-02/15-33-03

BFG run is complete! When ready, run: git reflog expire --expire=now --all && git gc --prune=now --aggressive
```

The repository should now be much smaller, and large files have been removed from history.
